### PR TITLE
Resolved IDENTITY-2558

### PR DESCRIPTION
**Pull request by IndunilRathnayake**

### CODE REVIEW

**Comment by IndunilRathnayake** [comment]
Issue: When creating a SP with the federation type as "Federated Authentication" in Local & Outbound Authentication Configuration, the selected federated IDP is saved with current default authenticator. If the default authenticator of IDP is changed afterwards, it is not reflected in SP side. So that when do authentication in SP application, it always authenticated through the saved default authenticator of federated IDP.
Fix: Added a listener to change the IDP's default authenticator in SP side, at the moment of changing IDP configurations.

Other changes:
1) added a new Exception in idp-mgt module (previously it uses "IdentityApplicationManagementException" defines in application-mgt)
2) added pre and post methods in ApplicationMgtListener and IdentityProviderMgtLister
3) Abstract listener for both IdentityProviderMgtLister & ApplicationMgtListener
4) Throw exception in ConfigurationFacade and catch and log in authentication framework 
5) Change return type of createApplication to void. (The int value returned is the auto-generated table ID which should not be returned, and isn't used anywhere in the client side as well.)

File: components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java [hunk 1]
```diff
@@ -89,19 +89,19 @@ public static IdentityProviderManager getInstance() {
      *
      * @param tenantDomain Tenant Domain
      * @return Tenant Id of domain user belongs to.
-     * @throws IdentityApplicationManagementException Error when getting tenant id from tenant
+     * @throws IdentityProviderManagementException Error when getting tenant id from tenant
      *                                                domain
      */
     private static int getTenantIdOfDomain(String tenantDomain)
```
Comment (johannnallathamby) [inline components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java]: Use method in IdentityUtil

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/internal/ApplicationMgtListenerServiceComponent.java [hunk 2]
```diff
@@ -30,27 +32,58 @@
  * interface="org.wso2.carbon.identity.application.mgt.listener.ApplicationMgtListener"
  * cardinality="0..n" policy="dynamic"
  * bind="setApplicationMgtListenerService"
- * unbind="unsetApplicationMgtListenerService" *
+ * unbind="unsetApplicationMgtListenerService"
  */
 public class ApplicationMgtListenerServiceComponent {
 
-    private static List<ApplicationMgtListener> listners = new ArrayList<ApplicationMgtListener>();
+    private static Map<Integer, ApplicationMgtListener> applicationMgtListeners;
```
Comment (johannnallathamby) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/internal/ApplicationMgtListenerServiceComponent.java]: Seems no use of the map here

Comment (IndunilRathnayake) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/internal/ApplicationMgtListenerServiceComponent.java]: Fixed in https://github.com/wso2/carbon-identity/pull/1146

**Pull request merged**